### PR TITLE
Improve `OnSystem` RuboCop

### PR DESCRIPTION
**Pull request by Rylan12**
This PR updates the existing `on_{macos,linux}` RuboCop checks to work for all `on_{system}` blocks. This checks for `if` statements in the top-level and correct them to `on_{system}`, and also checks for `on_{system}` blocks in `def install`, `test do`, and others to convert them to `if` statements.

`brew style --fix` should (ideally) be able to fix most of these issues on its own. The only time that it intentionally does not autocorrect (that I know of) is when there is an `else` statement after `if MacOS.version ...`. In that case, it's much harder to figure out what the correct `on_{system}` block for the `else` should be, so it just points out the error and doesn't correct.

I've also updated the tests to be more extensive and test the auto-correct ability.

Of course, this PR can't be merged until the style issues in our taps are fixed, and that can't happen until we have a new Homebrew/brew tag (which will happen soon once the new `on_{system}` blocks have been around for a bit longer).

### CODE REVIEW

**Review by MikeMcQuaid — APPROVED** [review]
Looks great!

File: Library/Homebrew/rubocops/components_order.rb
Comment (Rylan12) [inline Library/Homebrew/rubocops/components_order.rb:92]: `on_{system}` blocks in `resource` blocks can now contain variable assignments, `patch` blocks, and other `on_{system}` blocks.

File: Library/Homebrew/rubocops/components_order.rb
Comment (Rylan12) [inline Library/Homebrew/rubocops/components_order.rb:130]: `on_{system}` blocks can now contain `url`, `version`, `sha256`, `mirror`, and `fails_with` calls as well as other `on_{system}` blocks.

**Review by Rylan12 — COMMENTED** [review]
I've now also updated the components order cop to properly check the new `on_{system}` blocks. After fixing issues in homebrew/core, I found some instances where I felt that the current restrictions should be updated. I've commented these changes below so it's clear where the rules have _changed_ as opposed to just being extended to the new `on_{system}` methods.

**Comment by Rylan12** [comment]
Now that I think about it, we will probably have to separate these two cops into individual PRs. We need the updated component order cop _before_ making changes to formulae since it allows for the necessary methods to be used in the correct places. Then, once we've updated everything, we'll be able to enable the cop that requires that `on_{system}` is used instead of `if`. However, I'm going to sleep so that is something I'll do tomorrow.

File: Library/Homebrew/rubocops/components_order.rb
Comment (MikeMcQuaid) [reply Library/Homebrew/rubocops/components_order.rb:130]: > only raises issues with [libgccjit](https://github.com/Homebrew/homebrew-core/blob/master/Formula/libgccjit.rb) it turns out, and that formula definitely seems is a unique case.

@fxcoudert may have an insight on how long-lived this is likely to be.

> Thoughts on the best way to handle that formula?

Probably depends on @fxcoudert thoughts. I can see an argument that even if we only need it in this one case: some taps may need it in more e.g. those that ship binaries in their `url`.

Comment (Rylan12) [reply Library/Homebrew/rubocops/components_order.rb:130]: Okay, that makes sense. In the meantime, I think I'll hold off on allowing these and just temporarily not change `libgccjit`. That way, we can get a tag and start updating the rest of the formulae and casks, and then we'll just have to come back and make a decision before merging this PR (and banning `if Hardware::CPU.arm?`.

Comment (MikeMcQuaid) [reply Library/Homebrew/rubocops/components_order.rb:130]: Sounds good!

**Comment by Rylan12** [comment]
I've extracted the components order cop to https://github.com/Homebrew/brew/pull/13503 so it can be merged first.

**Comment by Rylan12** [comment]
Now that we have `on_system`, this will correct `if MacOS.version <= :monterey` to `on_system :linux, macos: :monterey_or_older`. It will still correct all other if conditions to the correct `on_{system}` methods. Also, it will now correct `on_system` to the appropriate `if` statement if it's used inside a forbidden place

**Comment by MikeMcQuaid** [comment]
🎉 great work here!

**Pull request merged**
